src/pages/signin/SignInPage.js

```diff
@@ -35,153 +32,163 @@ const propTypes = {
-        // - AND the login is not validated
-        const showUnlinkLoginForm =
-            Boolean(this.props.credentials.login && this.props.account.primaryLogin) && this.props.account.primaryLogin !== this.props.credentials.login && !this.props.account.validated;
-
-        // Show the old password form if
-        // - A login has been entered
-        // - AND an account exists and is validated for this login
-        // - AND a password hasn't been entered yet
-        // - AND haven't forgotten password
-        // - AND the login isn't an unvalidated secondary login
-        // - AND the user is NOT on the passwordless beta
-        const showPasswordForm =
-            Boolean(this.props.credentials.login) &&
-            this.props.account.validated &&
-            !this.props.credentials.password &&
-            !this.props.account.forgotPassword &&
-            !showUnlinkLoginForm &&
-            !Permissions.canUsePasswordlessLogins(this.props.betas);
-
-        // Show the new magic code / validate code form if
-        // - A login has been entered or a validateCode has been cached from sign in link
-        // - AND the login isn't an unvalidated secondary login
-        // - AND the user is on the 'passwordless' beta
-        const showValidateCodeForm =
-            Boolean(this.props.credentials.login || this.props.credentials.validateCode) && !showUnlinkLoginForm && Permissions.canUsePasswordlessLogins(this.props.betas);
-
-        // Show the resend validation link form if
-        // - A login has been entered
-        // - AND is not validated or password is forgotten
-        // - AND the login isn't an unvalidated secondary login
-        // - AND user is not on 'passwordless' beta
-        const showResendValidationForm =
-            Boolean(this.props.credentials.login) &&
-            (!this.props.account.validated || this.props.account.forgotPassword) &&
-            !showUnlinkLoginForm &&
-            !Permissions.canUsePasswordlessLogins(this.props.betas);
-
-        let welcomeHeader = '';
-        let welcomeText = '';
-        if (showValidateCodeForm) {
-            if (this.props.account.requiresTwoFactorAuth) {
+    }, []);
+
+    const {
+        shouldShowLoginForm,
+        shouldShowUnlinkLoginForm,
+        shouldShowPasswordForm,
+        shouldShowValidateCodeForm,
+        shouldShowResendValidationForm,
+        shouldShowWelcomeHeader,
+        shouldShowWelcomeText,
+    } = getRenderOptions({
+        hasLogin: Boolean(credentials.login),
+        hasPassword: Boolean(credentials.password),
+        hasValidateCode: Boolean(credentials.validateCode),
+        isPrimaryLogin: account.primaryLogin && account.primaryLogin === credentials.login,
+        isAccountValidated: Boolean(account.validated),
+        didForgetPassword: Boolean(account.forgotPassword),
+        canUsePasswordlessLogins,
+    });
+
+    const {welcomeHeader, welcomeText} = useMemo(() => {
```

Comment:
Maybe I'm misunderstanding your confusion, but maybe you can read up on how `React.useMemo` works a bit: https://react.dev/reference/react/useMemo

The function passed to `useMemo` returns an object with the `welcomeHeader` and `welcomeText`. However, that function is not free because it's doing some translation / regex replacing, etc... so by wrapping that logic in `useMemo`, the function used to compute the `welcomeHeader` and `welcomeText` is only executed when any of it's dependencies (the array passed as the second argument to `useMemo`) change.

Now, it's possible that `useMemo` isn't needed here because it's mostly a performance optimization. So by using it here I'm basically betting that the performance cost of computing `welcomeHeader` and `welcomeText` outweighs the cost of checking if any of the dependencies changed.

Reply:
Is this actually improving render performance for this page noticeably? Aren't `props.translate` calls basically free? There is formatting and a regex replace for SMS logins but I'm doubting that's expensive enough to warrant memoizing. I ask because this `useMemo` call is a lot more verbose than not using it.

Reply:
I don't necessarily agree with the readability argument against `useMemo` because I think it provides a nice way to encapsulate some of the component's logic in a way where it's really clear what it's doing.

If I have this:

```js
const {welcomeHeader, welcomeText} = useMemo(/* Big block of logic here */);
```

It's immediately obvious that whatever's in that `useMemo` call is being used to compute those two values.

```
let welcomeHeader;
let welcomeText;
// Big block of logic here
```

With this you have to actually read through the code to see what's used and why that code's there. It's not clear if there are side-effects in there, or what values are important.

Reply:
Answering the actual question of whether `useMemo` is good for performance is kind of beyond my skill level here / not worth digging into in this case because AFAICT the page works fine either way.

Reply:
Updated to remove the `useMemo` because this logic was in `render` before so not using `useMemo` is closer to the status quo